File: src/main/java/com/example/demo/config/LoginRateLimitFilter.java
package com.example.demo.config;

import com.example.demo.util.IpUtils;
import com.github.benmanes.caffeine.cache.Cache;
import com.github.benmanes.caffeine.cache.Caffeine;
import io.github.bucket4j.Bucket;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.extern.slf4j.Slf4j;
import org.jspecify.annotations.NonNull;
import org.springframework.core.annotation.Order;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import java.io.IOException;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.concurrent.TimeUnit;

/**
 * ログイン試行レート制限フィルター
 * ログインエンドポイントへのリクエスト数を制限してブルートフォース攻撃を防ぐ
 * 設定: 5分間に5回のログイン試行まで許可
 */
@Slf4j
@Component
@Order(2)
public class LoginRateLimitFilter extends OncePerRequestFilter {

    private final boolean trustProxy;
    private final long capacity;
    private final long refillMinutes;

    private final Cache<String, Bucket> cache;

    public LoginRateLimitFilter(AppProperties properties) {
        this.trustProxy = properties.isTrustProxy();
        this.capacity = properties.getLogin().getRateLimit().getCapacity();
        this.refillMinutes = properties.getLogin().getRateLimit().getRefillMinutes();
        this.cache = Caffeine.newBuilder()
                .maximumSize(10_000)
                .expireAfterAccess(Math.max(refillMinutes * 2, 10), TimeUnit.MINUTES)
                .build();
    }

    @Override
    protected void doFilterInternal(
            HttpServletRequest request,
            @NonNull HttpServletResponse response,
            @NonNull FilterChain filterChain) throws ServletException, IOException {
        String clientIp = IpUtils.clientIp(request, trustProxy);
        Bucket bucket = cache.get(clientIp, key -> createBucket());

        if (bucket.tryConsume(1)) {
            log.debug("ログイン試行許可: IP={}, 残り試行回数={}", clientIp, bucket.getAvailableTokens());
            filterChain.doFilter(request, response);
        } else {
            log.warn("ログイン試行レート制限超過: IP={}", clientIp);

            response.setStatus(HttpStatus.TOO_MANY_REQUESTS.value());
            response.setContentType("text/html; charset=UTF-8");
            response.setCharacterEncoding("UTF-8");
            String message = "ログイン試行回数が上限に達しました。" + refillMinutes + "分後に再度お試しください。";
            response.sendRedirect(
                    "/login?error=true&message=" + URLEncoder.encode(message, StandardCharsets.UTF_8));
        }
    }

    private Bucket createBucket() {
        return Bucket.builder()
                .addLimit(limit -> limit.capacity(capacity).refillGreedy(capacity, Duration.ofMinutes(refillMinutes)))
                .build();
    }

    @Override
    protected boolean shouldNotFilter(HttpServletRequest request) {
        return !("/login".equalsIgnoreCase(request.getRequestURI()) && "POST".equals(request.getMethod()));
    }

    @Override
    protected boolean shouldNotFilterAsyncDispatch() {
        return false;
    }

    @Override
    protected boolean shouldNotFilterErrorDispatch() {
        return false;
    }
}

File: src/main/kotlin/com/example/demo/config/LoginRateLimitFilter.kt
package com.example.demo.config

import com.example.demo.util.IpUtils
import io.github.bucket4j.Bucket
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.core.annotation.Order
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter
import java.time.Duration
import java.util.concurrent.ConcurrentHashMap

/**
 * ログイン試行レート制限フィルター
 * ログインエンドポイントへのリクエスト数を制限してブルートフォース攻撃を防ぐ
 * 設定: 5分間に5回のログイン試行まで許可
 */
@Component
@Order(2)
class LoginRateLimitFilter(
    @Value("\${app.trust-proxy:false}") private val trustProxy: Boolean,
    @Value("\${app.login.rate-limit.capacity:5}") private val capacity: Long,
    @Value("\${app.login.rate-limit.refill-minutes:5}") private val refillMinutes: Long
) : OncePerRequestFilter() {

    private val cache: MutableMap<String, Bucket> = ConcurrentHashMap()

    companion object {
        private val log = LoggerFactory.getLogger(LoginRateLimitFilter::class.java)
    }

    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {
        // ログインエンドポイント以外はスキップ
        if (!request.requestURI.equals("/login", ignoreCase = true) || request.method != "POST") {
            filterChain.doFilter(request, response)
            return
        }

        val clientIp = IpUtils.clientIp(request, trustProxy)
        val bucket = cache.computeIfAbsent(clientIp) { createBucket() }

        if (bucket.tryConsume(1)) {
            log.debug("ログイン試行許可: IP=$clientIp, 残り試行回数=${bucket.availableTokens}")
            filterChain.doFilter(request, response)
        } else {
            log.warn("ログイン試行レート制限超過: IP=$clientIp")

            response.apply {
                status = HttpStatus.TOO_MANY_REQUESTS.value()
                contentType = "text/html; charset=UTF-8"
                characterEncoding = "UTF-8"

                // ログイン画面にリダイレクトしてエラーメッセージを表示
                sendRedirect(
                    "/login?error=true&message=" +
                            java.net.URLEncoder.encode(
                                "ログイン試行回数が上限に達しました。${refillMinutes}分後に再度お試しください。",
                                "UTF-8"
                            )
                )
            }
        }
    }

    private fun createBucket(): Bucket =
        Bucket.builder()
            .addLimit { limit ->
                limit.capacity(capacity).refillGreedy(capacity, Duration.ofMinutes(refillMinutes))
            }
            .build()

    override fun shouldNotFilter(request: HttpServletRequest): Boolean {
        // POSTメソッドの/loginのみフィルタリング
        return !(request.requestURI.equals("/login", ignoreCase = true) && request.method == "POST")
    }

    override fun shouldNotFilterAsyncDispatch(): Boolean = false
    override fun shouldNotFilterErrorDispatch(): Boolean = false
}

